Refactored main and restructured code

ray-chew opened the pull request:
`__main__.py` is now cleaned, and all the initialisation and analysis steps have been moved to their respective modules or subpackages.

ray-chew committed:
freezing this version as the base comparison for the refactoring of the code

ray-chew committed:
checkpoint in restructuring and making the package more pythonic in structure

there are still a lot of broken imports and references, so the code does not run yet.

ray-chew committed:
done with restructuring the initialisation and preparation steps in __main__.py

still untested as I have to move the analysis steps to the DA subpackage before this will run

ray-chew committed:
moved blending from main() into its own submodule

also created a placeholder for the parallel integrator tests planned

ray-chew committed:
code runs until analysis step

the restructuring continues

ray-chew committed:
main has been restructured and all three dynamics tests passed

DA and blending are not covered in the tests, will look into them after refactoring the flow solver

ray-chew committed:
updated readme on how to run a single simulation

ray-chew committed:
updated changelog

ray-chew committed:
updated run.py to run rising bubble test case

since I removed all initial conditions except this

ray-chew committed:
Merge branch 'develop' into refactor

ray-chew committed:
refactored diagnostics.py a little

ray-chew committed:
cleaned up blending related code in the flow solver

ray-chew committed:
renamed the dycore subpackage to flow_solver for consistency

ray-chew committed:
updated data structures for a proper sol_ens and simulation integration time

ray-chew committed:
updated thermodynamics data structure.

ray-chew committed:
done refactoring the preparation and initialisation steps

this is with the new ensemble data structure; the flow solver runs; yet to be tested.

ray-chew committed:
updated analysis with new ensemble structure

ray-chew committed:
reinstated a few changes in time_update; updating mem with solution arrays at the end of the flow solver's looping

this is because the tests were failing with a discrepancy.

ray-chew committed:
updated docs from dycore to flow_solver

ray-chew committed:
updated changelog

ray-chew committed:
#6 diagnostics now compute either the sum or the norm as a target

ray-chew committed:
reinstated run.py with refactored package structure

ray-chew committed:
renamed test_dycore to test_flow_solver

ray-chew committed:
commented out validation attempt in user_data

validation step will be moved to dataclass structure

tbnc commented:
I will review. 

Are the docs checks supposed to succeed?

ray-chew committed:
working towards a first test case for the blending module

ray-chew commented:
> I will review.
> 
> Are the docs checks supposed to succeed?

Thank you. The checks broke due to the renaming of the subpackages. This issue has been resolved with the latest commits. But the pytests are broken now, as I am in the midst of restructuring the diagnostics module to accept more general definitions of test cases.

ray-chew committed:
flow solver tests now work with the new diagnostics data structure

more cleaning up is due for the diagnostics modules, but at least now the tests are no longer hardcoded. the blending has potential problems with failing tests, will look into this next.

tbnc commented on README.md:
````diff
@@ -74,7 +74,13 @@ A simple test can be found in [`run_scripts.test_dycore`](https://github.com/ray
-However, the codebase is structured such that the user can easily assemble a run script to define their own experiments. Refer to the documentation for the [available APIs](https://ray-chew.github.io/pyBELLA/apis.html).
+To run a simulation:
+```console
+pybella -ic rb -N 1
````
This does not appear to work, at least out of the box:

```
[nhad@ac6-101 pyBELLA]$ pybella -ic rb -N 1
Traceback (most recent call last):
  File "/home/nhad/.local/bin/pybella", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/etc/ecmwf/nfs/dh1_home_b/nhad/pyBELLA/src/__main__.py", line 30, in main
    sim_state = prepare.initialise()
                ^^^^^^^^^^^^^^^^^^^^
  File "/etc/ecmwf/nfs/dh1_home_b/nhad/pyBELLA/src/utils/prepare.py", line 43, in initialise
    ud.coriolis_strength = np.array(ud.coriolis_strength)
                                    ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'UserDataInit' object has no attribute 'coriolis_strength'
```

tbnc replied:
This does not appear to work, at least not out of the box:

```
[nhad@ac6-101 pyBELLA]$ pybella -ic rb -N 1
Traceback (most recent call last):
  File "/home/nhad/.local/bin/pybella", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/etc/ecmwf/nfs/dh1_home_b/nhad/pyBELLA/src/__main__.py", line 30, in main
    sim_state = prepare.initialise()
                ^^^^^^^^^^^^^^^^^^^^
  File "/etc/ecmwf/nfs/dh1_home_b/nhad/pyBELLA/src/utils/prepare.py", line 43, in initialise
    ud.coriolis_strength = np.array(ud.coriolis_strength)
                                    ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'UserDataInit' object has no attribute 'coriolis_strength'
```

tbnc commented on run_scripts/test_dycore.py:
```diff
@@ -1,16 +0,0 @@
-import pytest
```
Line 74 in the README https://github.com/ray-chew/pyBELLA/blob/2165bed4ee231e34f8ad460d46877585d44fc737/README.md?plain=1#L74 should take into account the fact that the file has been deleted.

tbnc commented on src/data_assimilation/analysis.py:
```diff
@@ -0,0 +1,123 @@
+import logging
```
With a view to the forthcoming published version, I would propose removing commented code lines sooner rather than later unless still part of prototyping (which probably should not end up in develop anyway?) or there is another very compelling reason to keep them in.

tbnc commented on src/flow_solver/discretisation/time_update.py:
```diff
@@ -57,16 +61,9 @@ def data_init(ud):
-    mpv,
-    t,
+    sst,
+    mem,
```
Nice and compact

ray-chew replied:
Thanks!

tbnc commented on pyproject.toml:
```diff
@@ -3,15 +3,15 @@ name = "pyBELLA"
-    "PyYAML==6.0",
-    "scipy==1.7.3",
-    "termcolor==2.4.0"
+    "dask",
```
2024.2.1

tbnc commented on pyproject.toml:
```diff
@@ -3,15 +3,15 @@ name = "pyBELLA"
 
 dependencies = [
-    "dask==2022.7.0",
-    "dill==0.3.6",
```
0.3.8

tbnc commented on pyproject.toml:
```diff
@@ -3,15 +3,15 @@ name = "pyBELLA"
 dependencies = [
-    "dask==2022.7.0",
-    "dill==0.3.6",
-    "h5py==3.7.0",
```
3.12.1

tbnc commented on pyproject.toml:
```diff
@@ -3,15 +3,15 @@ name = "pyBELLA"
-    "dask==2022.7.0",
-    "dill==0.3.6",
-    "h5py==3.7.0",
-    "matplotlib==3.5.1",
```
3.8.4

tbnc commented on pyproject.toml:
```diff
@@ -3,15 +3,15 @@ name = "pyBELLA"
-    "dill==0.3.6",
-    "h5py==3.7.0",
-    "matplotlib==3.5.1",
-    "numba==0.56.4",
```
0.60.0

tbnc commented on pyproject.toml:
```diff
@@ -3,15 +3,15 @@ name = "pyBELLA"
-    "h5py==3.7.0",
-    "matplotlib==3.5.1",
-    "numba==0.56.4",
-    "numpy==1.22.1",
```
1.26.4

tbnc commented on pyproject.toml:
```diff
@@ -3,15 +3,15 @@ name = "pyBELLA"
-    "matplotlib==3.5.1",
-    "numba==0.56.4",
-    "numpy==1.22.1",
-    "PyYAML==6.0",
```
6.0.2

tbnc commented on pyproject.toml:
```diff
@@ -3,15 +3,15 @@ name = "pyBELLA"
-    "numba==0.56.4",
-    "numpy==1.22.1",
-    "PyYAML==6.0",
-    "scipy==1.7.3",
```
1.14.1

tbnc commented on pyproject.toml:
```diff
@@ -3,15 +3,15 @@ name = "pyBELLA"
-    "numpy==1.22.1",
-    "PyYAML==6.0",
-    "scipy==1.7.3",
-    "termcolor==2.4.0"
```
2.4.0

ray-chew committed:
cleaned up blending module; do_blending_bef_timestep accepts mem data container

ray-chew committed:
fixed failing flow solver tests

I was not updating mem attributes in the time update while loop. This means that everytime the blending scheme sees mem, it resets the solution fields to the beginning of the assimilation window.

ray-chew committed:
blending routine runs with new data structures

I have not yet checked if the initial blending results are correct; will do this via the diagnostics module, but it does not work yet, as I have not yet generalised the target generation features.

ray-chew committed:
WIP: generalised generate target for diagnostics and reinstated plot_compare

nothing really works yet, because there seems to be a discrepancy between the target saved and the test values computed and compared against the target values. This has something to do with how the time steps are counted. Will need to carry on debugging.

ray-chew committed:
minor refactoring of utils.io and utils.prepare

ray-chew committed:
fixed bug with discrepancy between generated target and test values

the generalised target generation in the diagnostics module now works, the problem was with writing the output before the mem attributes were correctly updated.

ray-chew committed:
update all test ICs to support updating target; test_targets.yml updated with warm bubble test values

blending is still switched off, but this commit allows for blending to be tested as a next step. woohoo.

ray-chew committed:
added comments for DiagnosticState attributes

ray-chew committed:
extended diagnostics module to output more sensible plots

specifically, the primitive variables and the pressure perturbation. But the pressure perturbation computations are incorrect, and the background pressure distribution is still visible, I have no idea why...

ray-chew committed:
black formatted whole repo

ray-chew committed:
updated test diagnostics to compute L2 and max norms as errors, tolerances for each attribute type, and to output time increment of pressure fields for blending tests

ray-chew committed:
fixed pressure imprint bug #34

ray-chew committed:
latest yml targets

although this is now no longer used

ray-chew committed:
moved diagnostics tolerance check to the max abs error. all tests pass.

ray-chew committed:
fully separated the blending module (#3)

ray-chew committed:
created debug writer object

ray-chew committed:
removed deprecated functions to run with modern Python

see #33 for version of Python and dependencies

ray-chew committed:
updated requirements.txt

ray-chew committed:
first attempt at reinstating unstable lamb IC (#37)

ray-chew committed:
unstable lamb wave test (#37) is getting close to Chew et al. 2022

ray-chew committed:
fixed non-fatal bug in UserDataInit

ray-chew committed:
refactor io import run scripts

ray-chew committed:
fixed typo in EnsembleState.get_grid

ray-chew committed:
renamed LAMB_BDRY with ATMOSPHERIC_EXTENSION

ray-chew committed:
fixed typo in the unstable lamb test

so the initial 30 time steps or so are very close to the 0.4.0 release, but the solutions diverge after about 50 time steps due to build up in this difference and the sensitivity of the instabilities. However, the amplitudes are slightly larger in our current run and 0.4.0 slightly underestimated the instability growth rate. So we might actually be on the better side of things here. Nevertheless, may be worth looking into this in the future.

ray-chew committed:
rethought the flow solver tests

now that we have the unstable lamb wave, the actual lamb wave test is now a long 800-time-step run. The internal long wave and travelling vortex use the integrated hydrostatic module, and the rest of the test cases use the analytical hydrostatic state (exponentially decaying pressure). These four flow solver tests took me a minute to run on my laptop, and they produce pretty large output files, especially for the long lamb wave run. Possibly have to improve this in the future.

ray-chew committed:
extended hydrostatics module to integrate or provide analytical background

the former was the default and works for most cases, while the latter is necessary for well-balanced numerical tests.

ray-chew committed:
make the rising bubble test work with the intergrated hydrostatic module

ray-chew committed:
moved tests to a specific subdirectory

ray-chew committed:
moved rayleigh boundary handling to boundary.py (#36)

unstable test passed

ray-chew committed:
added a skeleton .coveragerc file

I am excluding the data assimilation module for now.

ray-chew committed:
updated workflow file for tests, coverage, and doc deployment

this resolves #28 and #29, and #38 partially.

ray-chew committed:
added stripped target outputs for CI tests

I can think of a better approach in the future

ray-chew committed:
updated test IC files to be consistent with target outputs

ray-chew committed:
fixed issues in deploy.yml

I hope the GitHub Actions work now

ray-chew committed:
added postprocessing interface method to strip target files

ray-chew committed:
this approach with pyproject.toml might work more generally

to be tested

ray-chew committed:
updated .gitignore

ray-chew committed:
updated requirements.txt to no longer install ancient dependencies

ray-chew committed:
cleaned up time_update.py

ray-chew committed:
overlooked a few issues that would cause CI to fail

ray-chew committed:
Using the default src layout for setuptools auto discovery

ray-chew committed:
reducing the number of timesteps and increasing tolerance for unstable lamb test

unstable lamb test is failing on CI as the numerical instability is architecture dependent. So maybe if we relax the checks, we can ensure that the instability develops similarly but also allow for numerical differences.

ray-chew committed:
extended diagnostics for the time increments in blending tests

ray-chew committed:
latest target file for the unstable lamb test

ray-chew committed:
reinstate all tests in deploy.yml

ray-chew committed:
added target file for the blending warm bubble test

ray-chew committed:
CI to also upload logs into artifact

ray-chew committed:
further increase tolerance for unstable lamb test

ray-chew committed:
removing docs compilation for now until refactoring is done

since I may change module names, paths, etc

ray-chew committed:
added commands for coverage and profiler to readme; updated gitignore
